### PR TITLE
Border StrokeDashArray

### PR DESCRIPTION
**Pull request by RomanSoloweow**
Adds properties to allow more flexible customization of Pen for Border
Base on #5843

### CODE REVIEW

**Comment by maxkatz6** [comment]
@Takoooooo why is it marked as a breaking change? I don't see any reason.

**Comment by Takoooooo** [comment]
> @Takoooooo why is it marked as a breaking change? I don't see any reason.

Sorry,missed that this helper class is internal

**Review by grokys — CHANGES_REQUESTED** [review]
A few comments on the code, but other than that I can see why being able to stroke borders with a dash pattern would be useful.

I'm not in love with the API, but this seems to be what we've inherited from WPF/UWP. Feels like a more sensible API would be to have a `BorderPen` property but I imagine WPF didn't go this direction because it makes binding more difficult in the common cases.

I guess `BorderBrush` and `BorderThickness` could be "aliases" for `BorderPen.Brush` and `BorderPen.Thickness` but that might raise complications?

I think we need to discuss this API a bit...

**Comment by grokys** [comment]
Actually, the existing border properties from `Pen` have the following name mapping:

`Pen.Brush` -> `BorderBrush`
`Pen.Thickness` -> `BorderThickness`

Given this, shouldn't these properties be called `BorderDashArray`, `BorderDashOffset` etc?

**Comment by grokys** [comment]
Actually, the existing border properties from `Pen` have the following name mapping:

`Pen.Brush` -> `BorderBrush`
`Pen.Thickness` -> `BorderThickness`

Given this, shouldn't these properties be called `BorderDashArray`, `BorderDashOffset` etc?

**Comment by grokys** [comment]
I can't see any effect in changing the stroke line cap currently; do line caps even make sense for borders?

File: src/Avalonia.Controls/Border.cs [hunk 1]
```diff
@@ -41,7 +52,31 @@ public partial class Border : Decorator, IVisualWithRoundRectClip
         /// </summary>
         public static readonly StyledProperty<BoxShadows> BoxShadowProperty =
             AvaloniaProperty.Register<Border, BoxShadows>(nameof(BoxShadow));
-        
+
+        /// <summary>
+        /// Defines the <see cref="StrokeDashOffset"/> property.
+        /// </summary>
+        public static readonly StyledProperty<double> StrokeDashOffsetProperty =
+            AvaloniaProperty.Register<Shape, double>(nameof(StrokeDashOffset));
```
[the remark this answers is not in the document]
Comment (RomanSoloweow) [reply src/Avalonia.Controls/Border.cs:60]: > I'm not in love with the API, but this seems to be what we've inherited from WPF/UWP.

Yes, many things in Avalonia are inherited from WPF. And I think we've already got to the point where we can ditch WPF aside and improve usability.

The lack of a StrokeDashArray in the Border is part of this legacy.
With the current implementation, Dotted Border required good preparation and was unbearable for newbies (I went through this myself).

> I guess BorderBrush and BorderThickness could be "aliases" for BorderPen.Brush and BorderPen.Thickness but that might raise complications?

I was relying on [Shape](https://github.com/AvaloniaUI/Avalonia/blob/8f8393978e9c9c42e5ee855f001a39a23be35589/src/Avalonia.Controls/Shapes/Shape.cs) when adding these properties. I think the unification in BorderPen will cause some confusion.
or it is necessary to change this in Shape too.

**Comment by RomanSoloweow** [comment]
@grokys 

> I can't see any effect in changing the stroke line cap currently; do line caps even make sense for borders?
What do you have in mind? Effect behaves as expected
image

File: src/Avalonia.Controls/Utils/BorderRenderHelper.cs [hunk 2]
```diff
@@ -15,8 +17,13 @@ internal class BorderRenderHelper
         private Size _size;
         private Thickness _borderThickness;
         private CornerRadius _cornerRadius;
+        private AvaloniaList<double> _strokeDashArray;
+        private double _strokeDashOffset;
+        private PenLineCap _strokeLineCap;
+        private PenLineJoin _strokeJoin;
```
[the remark this answers is not in the document]
Comment (RomanSoloweow) [reply src/Avalonia.Controls/Utils/BorderRenderHelper.cs]: If we combine them into one inner class, changing one of the properties should create the class, and therefore all pointers.

Or we can ignore all of these properties (not create an inner class) until the strokeDashArray is exposed. After all, without it, these properties have no effect.

[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Controls/Utils/BorderRenderHelper.cs]: Yes, exactly. We'd create an instance of the inner class once one of the properties was required, so borders that don't need StrokeDashArray's would only grow by 1 pointer.

**Comment by grokys** [comment]
Looks good to me I think, other than moving the uncommon properties to a separate class in `BorderRenderHelper`. @Gillibald @maxkatz6 @wieslawsoltes any input?

**Comment by RomanSoloweow** [comment]
@grokys on @maxkatz6  advice, launched benchmarkets 8

Current master:
image

This pool request
image

File: src/Avalonia.Controls/Utils/BorderRenderHelper.cs [hunk 3]
```diff
@@ -84,17 +92,26 @@ void Update(Size finalSize, Thickness borderThickness, CornerRadius cornerRadius
...
             Size finalSize, Thickness borderThickness, CornerRadius cornerRadius,
-            IBrush background, IBrush borderBrush, BoxShadows boxShadows)
+            IBrush background, IBrush borderBrush, BoxShadows boxShadows, double borderDashOffset = 0,
+            PenLineCap borderLineCap = PenLineCap.Flat, PenLineJoin borderLineJoin = PenLineJoin.Miter,
+            AvaloniaList<double> borderDashArray = null)
         {
             if (_size != finalSize
                 || _borderThickness != borderThickness
                 || _cornerRadius != cornerRadius
+                || _borderLineCap != borderLineCap
```
Comment (Gillibald) [inline src/Avalonia.Controls/Utils/BorderRenderHelper.cs:102]: We don't need to update the geometry etc. when those new properties change. They don't affect measurements. Please remove them from this check.

Comment (Gillibald) [reply src/Avalonia.Controls/Utils/BorderRenderHelper.cs:102]: We always pass current values to RenderCore so the render path should always use updated values.

Comment (RomanSoloweow) [reply src/Avalonia.Controls/Utils/BorderRenderHelper.cs]: Yes you are right, thank you

**Comment by Gillibald** [comment]
Please rerun the benchmarks after your latest changes. Ideally, nothing should change now compared to master.

**Comment by RomanSoloweow** [comment]
@Gillibald my computer does not give stable results, but here

Current master:
image

This pool request
image

**Comment by RomanSoloweow** [comment]
@Gillibald @maxkatz6  @grokys in previous tests there was a different version of .Net Core. Here are the correct results:

Current master:
image


This pool request
image

I think now we can finally merge this PR :)

**Review by maxkatz6 — APPROVED** [review]

**Comment by RomanSoloweow** [comment]
@grokys

**Comment by grokys** [comment]
@RomanSoloweow we've got a problem with our CI failing at the moment due to MS I believe removing `Microsoft.DotNet.ApiCompat` from nuget, which we depended on. We can merge once we've sorted that mess out.

File: src/Avalonia.Controls/Utils/BorderRenderHelper.cs [hunk 4]
```diff
@@ -15,8 +17,13 @@ internal class BorderRenderHelper
         private Size _size;
         private Thickness _borderThickness;
         private CornerRadius _cornerRadius;
+        private AvaloniaList<double> _borderDashArray;
```
Comment (grokys) [inline src/Avalonia.Controls/Utils/BorderRenderHelper.cs:20]: I'm seeing a warning that these fields are unused on `master` - should they have been removed?

Comment (Gillibald) [reply src/Avalonia.Controls/Utils/BorderRenderHelper.cs:20]: Y they are not used anymore

Comment (RomanSoloweow) [reply src/Avalonia.Controls/Utils/BorderRenderHelper.cs:20]: Yes, this is my mistake, I didn't notice. But this is great news, as it means that the problem of new properties has been solved :)

now I will do a PR with a fix

**Pull request merged**